diff --git a/xbmc/addons/AddonDatabase.cpp b/xbmc/addons/AddonDatabase.cpp
index 8b21da88d6..a81b5ce24e 100644
--- a/xbmc/addons/AddonDatabase.cpp
+++ b/xbmc/addons/AddonDatabase.cpp
@@ -418,7 +418,7 @@ void CAddonDatabase::SyncInstalled(const std::set<std::string, std::less<>>& ids
     BeginTransaction();
     for (const auto& id : added)
     {
-      int enable = 0;
+      int enable = 1;
 
       if (system.contains(id) || optional.contains(id))
         enable = 1;
